benchmarktests/target_secret_totp.go

```diff
@@ -0,0 +1,171 @@
+}
+
+type TOTPTest struct {
+	pathPrefix string
+	header     http.Header
+	config     *TOTPSecretTestConfig
+	numkeys    int
+	keysize    int
+	action     string
+	logger     hclog.Logger
+}
+
+type TOTPSecretTestConfig struct {
+	NumKeys int `hcl:"numkeys,optional"`
+	KeySize int `hcl:"keysize,optional"`
+}
+
+func (a *TOTPTest) ParseConfig(body hcl.Body) error {
+	testConfig := &struct {
+		Config *TOTPSecretTestConfig `hcl:"config,block"`
+	}{
+		Config: &TOTPSecretTestConfig{
+			KeySize: 20,
+			NumKeys: 1000,
+		},
+	}
+
+	diags := gohcl.DecodeBody(body, nil, testConfig)
+	if diags.HasErrors() {
+		return fmt.Errorf("error decoding to struct: %v", diags)
+	}
+	a.config = testConfig.Config
+
+	return nil
+}
+
+func (t *TOTPTest) read(client *api.Client) vegeta.Target {
+	secnum := int(1 + rand.Int31n(int32(t.numkeys)))
+	return vegeta.Target{
+		Method: TOTPReadTestMethod,
+		URL:    client.Address() + t.pathPrefix + "/code/" + strconv.Itoa(secnum),
+		Header: t.header,
+	}
+}
+
+func (t *TOTPTest) write(client *api.Client) vegeta.Target {
+	secnum := int(1 + rand.Int31n(int32(t.numkeys)))
+	return vegeta.Target{
+		Method: TOTPWriteTestMethod,
+		URL:    client.Address() + t.pathPrefix + "/keys/write-test-" + strconv.Itoa(secnum),
+		Header: t.header,
+		Body:   []byte(`{"name":"` + strconv.Itoa(secnum) + `","generate":true,"issuer":"Vault","account_name":"test@example.com"}`),
+	}
+}
+
+func (t *TOTPTest) Target(client *api.Client) vegeta.Target {
+	switch t.action {
+	case "write":
+		return t.write(client)
+	default:
+		return t.read(client)
+	}
+}
+
+func (t *TOTPTest) Cleanup(client *api.Client) error {
+	t.logger.Trace(cleanupLogMessage(t.pathPrefix))
+	_, err := client.Logical().Delete(strings.Replace(t.pathPrefix, "/v1/", "/sys/mounts/", 1))
+	if err != nil {
+		return fmt.Errorf("error cleaning up mount: %v", err)
+	}
+	return nil
+}
+
+func (t *TOTPTest) GetTargetInfo() TargetInfo {
+	var method string
+	switch t.action {
+	case "write":
+		method = TOTPWriteTestMethod
+	default:
+		method = TOTPReadTestMethod
+	}
+	return TargetInfo{
+		method:     method,
+		pathPrefix: t.pathPrefix,
+	}
+}
+
+func (t *TOTPTest) Setup(client *api.Client, mountName string, topLevelConfig *TopLevelTargetConfig) (BenchmarkBuilder, error) {
+	var err error
+	secretPath := mountName
+
+	switch t.action {
+	case "write":
+		t.logger = targetLogger.Named(TOTPWriteTestType)
+	default:
+		t.logger = targetLogger.Named(TOTPReadTestType)
+	}
+
+	if topLevelConfig.RandomMounts {
+		secretPath, err = uuid.GenerateUUID()
+		if err != nil {
+			log.Fatalf("can't create UUID")
+		}
+	}
+
+	t.logger.Trace(mountLogMessage("secrets", "totp", secretPath))
+	err = client.Sys().Mount(secretPath, &api.MountInput{
+		Type: "totp",
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error mounting totp secrets engine: %v", err)
+	}
+
+	secval := map[string]interface{}{
+		"generate":     true,
+		"issuer":       "Vault",
+		"account_name": "test@example.com",
+	}
```

Comment:
These configuration options should be parameterized based on configuration from previous comment. We would also want to have these defaults be included as part of the configuration rather than in the setup method. The data can be extracted from the configuration struct using the `structToMap` function.